### ISSUE TITLE
Difference between Pyrodigal 2.0.3 and Pyrodigal 2.6.3

### ISSUE THREAD

**Issue by pchaumeil** [althonos/pyrodigal#27]
Hello,

Thanks for developing this great tool!

I have been running pyrodigal and prodigal in parallel against multiple NCBI genomes and I have some questions about the results.

if we take the example of [GCA_934162975.1](https://www.ncbi.nlm.nih.gov/assembly/GCA_934162975.1/)
I downloaded this genome then ran both prodigal and pyrodigal on it:

```
 prodigal -i GCA_934162975.1_ERR7745755_bin.57_genomic.fna -m -g 11 -o GCA_934162975.1_run/prodigal_2.6.3_features.gff -a GCA_934162975.1_run/prodigal_2.6.3_amino.faa -d GCA_934162975.1_run/prodigal_2.6.3_nuc.fna
 pyrodigal -i GCA_934162975.1_ERR7745755_bin.57_genomic.fna -m -g 11 -o GCA_934162975.1_run/pyrodigal_2.0.3_features.gff -a GCA_934162975.1_run/pyrodigal_2.0.3_amino.faa -d GCA_934162975.1_run/pyrodigal_2.0.3_nuc.fna
```

Pyrodigal version is Pyrodigal **v2.0.3** and Prodigal version is **2.6.3 with the latest patch**
when I compare the 2 results of these 2 commands:

```
cd GCA_934162975.1_run/
diff prodigal_2.6.3_amino.faa pyrodigal_2.0.3_amino.faa
3677c3677
< >CAKPJW010000121.1_3 # 2300 # 2467 # -1 # ID=121_3;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.107
---
> >CAKPJW010000121.1_3 # 2300 # 2467 # -1 # ID=121_3;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.113
4498,4501c4498,4502
< >CAKPJW010000172.1_3 # 1334 # 1873 # -1 # ID=172_3;partial=00;start_type=ATG;rbs_motif=AGGA;rbs_spacer=5-10bp;gc_cont=0.230
< MNVFEIILLSFGLAADAFATSICKGLELPRVTFKHMIIVGLFFGIFQGLMPLIGYFIGDS
< FSLYTVKFNNLIAFILLSIIGINMIHESFSKDEIYDNSFSFKNLFVLAFATSIDAFTSGI
< TFSLLDINIPIAIFFITIITFITSVLGVKIGSIFGYKYKSKAELLGGCILILLGIKLFV*
---
> >CAKPJW010000172.1_3 # 1334 # 1882 # -1 # ID=172_3;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.231
> GYLMNVFEIILLSFGLAADAFATSICKGLELPRVTFKHMIIVGLFFGIFQGLMPLIGYFI
> GDSFSLYTVKFNNLIAFILLSIIGINMIHESFSKDEIYDNSFSFKNLFVLAFATSIDAFT
> SGITFSLLDINIPIAIFFITIITFITSVLGVKIGSIFGYKYKSKAELLGGCILILLGIKL
> FV*
4567c4568
< >CAKPJW010000177.1_2 # 1419 # 1832 # -1 # ID=177_2;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.283
---
> >CAKPJW010000177.1_2 # 1419 # 1832 # -1 # ID=177_2;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.287
```

there are changes on 3 different headers(CAKPJW010000121.1_3, CAKPJW010000172.1_3, CAKPJW010000177.1_2) but only one gene is different (CAKPJW010000172.1_3).
If we look at the first header CAKPJW010000121.1_3, the difference in the header is about the gc_cont of the gene. so I had a look at the nucleotide sequence for this gene

```
grep CAKPJW010000121.1_3 -A 3 prodigal_2.6.3_nuc.fna
>CAKPJW010000121.1_3 # 2300 # 2467 # -1 # ID=121_3;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.107
ATGAAATATAATAAATTAAAAGATAAAACAGTAGAAGAATTAAAAGAAGAAATTAATAAAATAGATATTA
AAGTAAGTAAAAAAAGTAAGAAGTTTATTATAATAATTTTAATTTTTATAATATTAATTAGTATTATATT
ACTCATAATGTTAAATTTATTTAAATAA
grep CAKPJW010000121.1_3 -A 3 pyrodigal_2.0.3_nuc.fna
>CAKPJW010000121.1_3 # 2300 # 2467 # -1 # ID=121_3;partial=01;start_type=Edge;rbs_motif=None;rbs_spacer=None;gc_cont=0.113
ATGAAATATAATAAATTAAAAGATAAAACAGTAGAAGAATTAAAAGAAGAAATTAATAAAATAGATATTA
AAGTAAGTAAAAAAAGTAAGAAGTTTATTATAATAATTTTAATTTTTATAATATTAATTAGTATTATATT
ACTCATAATGTTAAATTTATTTAAATAA

```
Both sequences are identical but the GC content is different. from what I can see, Pyrodigal is returning 0.113 but it should be 0.107, is this correct?

Now when I look at the gene CAKPJW010000172.1_3, the 2 sequences are different between prodigal and pyrodigal. My understanding is that pryodigal and prodigal ( with the latest patch) should have the same results. Am I missing something here?

Thanks for your help, 
Let me know if you need anymore information.

**Comment by althonos** [comment]
The differences you were getting were actually coming from two different bugs, both of which I patched: 
- The GC content was not computed correctly for genes on the reverse strand starting on the edge; this was to fix an issue with Prodigal occasionally reading sequence data out of the boundaries of the allocated buffer. 
- The RBS score was not computed at all for genes on the edge, even when there was a chance for a partial match. This is why Pyrodigal made `CAKPJW010000172.1_3` start at the edge instead of at an `ATG` codon, as it didn't find the partial RBS motif, and thus discarded the `ATG` codon.

**Comment by althonos** [comment]
Fixed in `v2.0.4`.

**Comment by pchaumeil** [comment]
Awesome! Thanks for the quick fix!
